nimble/controller/src/ble_ll_sync.c

```diff
@@ -0,0 +1,1449 @@
+    }
+
+    sync_pending = false;
+    ble_ll_sync_sm_establishing = NULL;
+
+    for (i = 0; i < BLE_LL_SYNC_CNT; i++) {
+        sm = &ble_ll_sync_sm[i];
+
+        sm->flags &= ~ BLE_LL_SYNC_SM_FLAG_PENDING;
+
+        if (sm->flags & BLE_LL_SYNC_SM_FLAG_ESTABLISHING) {
+            ble_ll_sched_rmv_elem(&sm->sch);
+            ble_npl_eventq_remove(&g_ble_ll_data.ll_evq, &sm->sync_ev_end);
+        }
+    }
+
+    *post_cmd_cb = ble_ll_sync_cancel_complete_event;
+
+    return BLE_ERR_SUCCESS;
+}
+
+int
+ble_ll_sync_terminate(uint8_t *cmdbuf)
+{
+    struct ble_ll_sync_sm *sm;
+    uint16_t handle;
+
+    if (sync_pending) {
+        return BLE_ERR_CMD_DISALLOWED;
+    }
+
+    handle  = get_le16(cmdbuf);
+    if (handle > 0xeff) {
+        return BLE_ERR_INV_HCI_CMD_PARMS;
+    }
+
+    if (handle >= BLE_LL_SYNC_CNT) {
+        return BLE_ERR_UNK_ADV_INDENT;
+    }
+
+    sm = &ble_ll_sync_sm[handle];
+    if (!(sm->flags & BLE_LL_SYNC_SM_FLAG_ESTABLISHED)) {
+        return BLE_ERR_UNK_ADV_INDENT;
+    }
+
+    ble_ll_sync_sm_clear(sm);
+
+    return BLE_ERR_SUCCESS;
+}
+
+int
+ble_ll_sync_list_add(uint8_t *cmdbuf)
+{
+    struct ble_ll_sync_sm *sm;
+    uint8_t addr_type;
+    uint8_t *addr;
+    uint8_t sid;
+
+    if (sync_pending) {
+        return BLE_ERR_CMD_DISALLOWED;
+    }
+
+    addr_type = cmdbuf[0];
+    if (addr_type > BLE_HCI_ADV_PEER_ADDR_MAX) {
+        return BLE_ERR_INV_HCI_CMD_PARMS;
+    }
+
+    addr = &cmdbuf[1];
+
+    sid = cmdbuf[7];
+    if (sid > 0x0f) {
+        return BLE_ERR_INV_HCI_CMD_PARMS;
+    }
+
+    sm = ble_ll_sync_get(addr, addr_type, sid);
+    if (!sm) {
+        return BLE_ERR_MEM_CAPACITY;
+    }
+
+    sm->flags |= BLE_LL_SYNC_SM_FLAG_ON_LIST;
+
+    return BLE_ERR_SUCCESS;
+}
+
+int
+ble_ll_sync_list_remove(uint8_t *cmdbuf)
+{
+    struct ble_ll_sync_sm *sm;
+    uint8_t addr_type;
+    uint8_t *addr;
+    uint8_t sid;
+
+    if (sync_pending) {
+        return BLE_ERR_CMD_DISALLOWED;
+    }
+
+    addr_type = cmdbuf[0];
+    if (addr_type > BLE_HCI_ADV_PEER_ADDR_MAX) {
+        return BLE_ERR_INV_HCI_CMD_PARMS;
+    }
+
+    addr = &cmdbuf[1];
+
+    sid = cmdbuf[7];
+    if (sid > 0x0f) {
+        return BLE_ERR_INV_HCI_CMD_PARMS;
+    }
+
+    sm = ble_ll_sync_find(addr, addr_type, sid);
+    if (!sm) {
+        return BLE_ERR_UNK_ADV_INDENT;
+    }
+
+    /* if sync is established only mark entry as removed from list */
+    if (sm->flags & BLE_LL_SYNC_SM_FLAG_ESTABLISHED) {
+        sm->flags &= ~BLE_LL_SYNC_SM_FLAG_ON_LIST;
+    } else {
+        ble_ll_sync_sm_clear(sm);
+    }
+
+    return BLE_ERR_SUCCESS;
+}
+
+int
+ble_ll_sync_list_clear(void)
+{
+    struct ble_ll_sync_sm *sm;
+    int i;
+
+    if (sync_pending) {
+        return BLE_ERR_CMD_DISALLOWED;
+    }
+
+    for (i = 0; i < BLE_LL_SYNC_CNT; i++) {
+        sm = &ble_ll_sync_sm[i];
+
+        /* if sync is establish only mark entry as removed from list */
+        if (sm->flags & (BLE_LL_SYNC_SM_FLAG_ESTABLISHING |
```

Comment:
could it ever be `BLE_LL_SYNC_SM_FLAG_ESTABLISHING` if `sync_pending` is not set?